### PR TITLE
[MODULAR] Fixes bugs with meson aviators and all of the other modular glasses

### LINKED ISSUE #15628: Retinal Projector are permanent (and dont toggle)

### LINKED ISSUE #16920: Security HUD Aviators might cause permanent HUD view
Reporting client version: 514.1589

## Round ID:
6761

## Testmerges:
- [hammers dont damage, reagent weapons when imbued lose durability](https://github.com/Skyrat-SS13/Skyrat-tg/pull/16732)
- [Space hotel Tender loving Care (and more)](https://github.com/Skyrat-SS13/Skyrat-tg/pull/16290)
- [Ghost Hotel Update](https://github.com/Skyrat-SS13/Skyrat-tg/pull/16447)
- [Granite's Love pass 3 of X for ruins - DS-2 flavor](https://github.com/Skyrat-SS13/Skyrat-tg/pull/16769)

## Reproduction:
Equip Security HUD Aviators and use the 'Toggle Mode' button. After removing the HUD, the screen will still show an active HUD display and you will be able to view and set records by examining people. I managed to fix it once by wearing them, toggling the mode, and taking them off. When I did it a second time, the HUD was permanent.

image

### LINKED ISSUE #17682: Meson HUD Aviators lose their Hallucination protection when toggled
## Reproduction:
Spawn a pair of Meson aviators, check variables.

image

So far so good. Now turn them off and back on again.

image

I've experienced this bug before by accidentally toggling them and it's rather inconvenient. Affects both prescription and non prescription variants.

Prescription meson hud aviators are also missing from the integrated hud research node and thus the fabricator but that's probably beyond the scope of this report.

Thank you in advanced for whoever takes a look into this.

### LINKED ISSUE #18379: Welding mask + Inactive Retinal projector Meson HUD makes you blind temporarily
## Testmerges:
[#17455: 'Fairly large assault operatives changes' by Paxilmaniac at commit d14745ce56](https://github.com/Skyrat-SS13/Skyrat-tg/pull/17455)
[#18292: 'Hemophagus, The Hemophage Revamp, Part Three: A Taste For Blood' by GoldenAlpharex at commit c9cfe2a9e0](https://github.com/Skyrat-SS13/Skyrat-tg/pull/18292)

## Reproduction: 
Put on a welding helmet that is angled down, next get either retinal projector mesons or aviator mesons, turn them off and put them on

### PR DESCRIPTION
**Pull request by vinylspiders**
## About The Pull Request

Fixes #17682 which affected all modular glasses, not just the mesons.
Fixes #16920
Fixes #15628
Fixes #18379

The bugs were too numerous to really list them all. It was a port that wasn't ported right in the first place (or perhaps the original code was also bugged? hard to say) so I ended up just redoing a lot of the code from scratch. No changes to how the items are supposed to work in game, though. The original intent has been preserved. Just fixed the code to not be a nightmare.

A lot of problems were caused by the `initial()` proc being used incorrectly. `initial()` determines the initial value at compile time, so it was setting them back to the parent type's values and not the subtypes' which was the initial (seewhatididthere) coder's intent. On top of that one of the fields was a list, which initial actually does not work on in this version of byond.

A lot of the other bugs were from failing to do certain things, e.g. the component that tells the worn sprite to update when update_appearance gets called on the object was not set correctly. HUDs were not being added/removed properly. All of that's been fixed.

The logic of the toggle switch block has been cleaned up because it was previously a nightmare to work with and very prone to breaking. It is now more resilient, and if any bugs should arise in the future they will be a whole lot easier to fix now.

tl;dr Fixed it so the toggle no longer ruins everything. 

## How This Contributes To The Skyrat Roleplay Experience

All the modular aviators were broken. Now they're not.

## Proof of Testing

<details>
<summary>Glasses are now working again</summary>

Security HUD aviators (prescription version)

image

image

Medical HUD aviators (prescription version)

image

image

Meson aviators

Initial

image

After turning off

image

After turning back on (traits are back)

image

Medical retinal projector

https://user-images.githubusercontent.com/13398309/215185987-b0f6b6a4-1dc4-4678-a003-8d289b2e06fe.mp4

image

image

The other retinal projectors are the same thing, but with different color schemes.

Welding works fine now, no blindness.

image

</details>

## Changelog

:cl:
fix: meson aviators and other modular glasses will no longer lose their qualities when toggled off and then back on
fix: hud aviators and projectors no longer can give you a permanent hud that cannot be removed even when the glasses are taken off
fix: meson aviators and retinal projectors no longer cause blindness while welding with mask on
/:cl:

### CODE REVIEW

File: modular_skyrat/modules/modular_items/code/modular_glasses.dm [hunk 1]
```diff
@@ -15,66 +15,127 @@
...
 	var/off_state = "salesman_fzz"
 	/// Sets a list of modes to cycle through
-	var/list/modes = list(MODE_OFF = MODE_ON, MODE_ON = MODE_OFF)
-	/// Defines initial mode upon initialisation, best to leave this on MODE_ON
-	var/mode = MODE_ON
+	var/list/modes = list(MODE_OFF, MODE_ON)
+	/// The current operating mode
+	var/mode
 	/// Defines messages that will be shown to the user upon switching modes (e.g. turning it on)
 	var/list/modes_msg = list(MODE_ON = "You activate the optical matrix on the ", MODE_OFF = "You deactivate the optical matrix on the ")
```
Comment (tf-4) [inline modular_skyrat/modules/modular_items/code/modular_glasses.dm]: just remove this and have the balloon alerts be "enabled"/"disabled" respectively, these are awful for balloon alerts

Comment (vinylspiders) [reply modular_skyrat/modules/modular_items/code/modular_glasses.dm]: Agreed, that is probably the one thing I would have changed functionally from the original. Is that better, or should I make them even shorter?

**Comment by Laria-D2** [comment]
I've noticed the prescription science glasses don't spawn with the "fixes_nearsighted" tag anymore. 
(Figured I'd mention it here instead of in a new issue report since it 'might be?' within this PR scope, but I'll gladly open one if that's what I'm supposed to do.)

**Comment by vinylspiders** [comment]
> I've noticed the prescription science glasses don't spawn with the "fixes_nearsighted" tag anymore. (Figured I'd mention it here instead of in a new issue report since it 'might be?' within this PR scope, but I'll gladly open one if that's what I'm supposed to do.)

Ah looks like there was a very recent refactoring of that. I updated this with the changes. Thanks for the heads up.

File: modular_skyrat/modules/modular_items/code/modular_glasses.dm [hunk 2]
```diff
@@ -15,66 +15,127 @@
...
 	/// Sets a list of modes to cycle through
-	var/list/modes = list(MODE_OFF = MODE_ON, MODE_ON = MODE_OFF)
-	/// Defines initial mode upon initialisation, best to leave this on MODE_ON
-	var/mode = MODE_ON
+	var/list/modes = list(MODE_OFF, MODE_ON)
+	/// The current operating mode
+	var/mode
 	/// Defines messages that will be shown to the user upon switching modes (e.g. turning it on)
-	var/list/modes_msg = list(MODE_ON = "You activate the optical matrix on the ", MODE_OFF = "You deactivate the optical matrix on the ")
+	var/list/modes_msg = list(MODE_ON = "Optical matrix enabled", MODE_OFF = "Optical matrix disabled")
```
Comment (tf-4) [inline modular_skyrat/modules/modular_items/code/modular_glasses.dm]:
```suggestion
	var/list/modes_msg = list(MODE_ON = "optical matrix enabled", MODE_OFF = "optical matrix disabled")
```
balloon alerts should always be lower case

Comment (vinylspiders) [reply modular_skyrat/modules/modular_items/code/modular_glasses.dm]: Got it

**Review by tf-4 — APPROVED** [review]
